src/v/redpanda/application.cc

```diff
@@ -131,6 +131,65 @@ application::application(ss::sstring logger_name)
 
 application::~application() = default;
 
+void application::shutdown() {
+    // Stop accepting new requests.
+    if (_kafka_server.local_is_initialized()) {
+        _kafka_server.invoke_on_all(&net::server::shutdown_input).get();
+    }
+    if (_rpc.local_is_initialized()) {
+        _rpc.invoke_on_all(&net::server::shutdown_input).get();
+    }
+
+    // We schedule shutting down controller input and aborting its operation as
+    // one of the first shutdown steps. This way we terminate all long running
+    // operations before shutting down the RPC server, preventing it from
+    // waiting on background dispatch gate `close` call.
+    if (controller) {
+        controller->shutdown_input().get();
+    }
+    if (kafka_group_migration) {
+        kafka_group_migration->await().get();
+    }
+
+    // Stop processing heartbeats before stopping the partition manager (and
+    // the underlying Raft consensus instances). Otherwise we'd process
+    // heartbeats for consensus objects that no longer exist.
+    if (raft_group_manager.local_is_initialized()) {
+        raft_group_manager.invoke_on_all(&raft::group_manager::stop_heartbeats)
+          .get();
+    }
+    // Stop all partitions before destructing the subsystems (transaction
+    // coordinator, etc). This interrupts ongoing replication requests,
+    // allowing higher level state machines to shutdown cleanly.
+    if (partition_manager.local_is_initialized()) {
+        partition_manager
+          .invoke_on_all(&cluster::partition_manager::stop_partitions)
+          .get();
+    }
+    if (cp_partition_manager.local_is_initialized()) {
+        cp_partition_manager
+          .invoke_on_all(&coproc::partition_manager::stop_partitions)
+          .get();
+    }
+
+    // Wait for all requests to finish before destructing services that may be
+    // used by pending requests.
+    if (_kafka_server.local_is_initialized()) {
+        _kafka_server.invoke_on_all(&net::server::wait_for_shutdown).get();
+        _kafka_server.stop().get();
+        _kafka_conn_quotas.stop().get();
+    }
+    if (_rpc.local_is_initialized()) {
+        _rpc.invoke_on_all(&net::server::wait_for_shutdown).get();
+        _rpc.stop().get();
+    }
+
+    // Shut down services in reverse order to which they were registered.
```

Comment:
nit: might be good to have a comment here explaining the rationale for shutting down some services explicitly in this function vs. letting some other services go via the generic _deferred mechanism, so that people adding things know whether they should shut them down here or use deferred.

Reply:
I will have some other PRs soon in this area -- I'll lump a comment into one of them.

I'll merge this now though so it has more bake time, and we can begin depending on it in other PRs.